### ISSUE TITLE
wire: generate body of factory methods

### ISSUE THREAD

**Issue by zhengyangfeng00** [google/go-cloud#690]
### Is your feature request related to a problem? Please describe.

I am trying to figure out how to use wire in factory pattern. For example I have a factory for creating object X which depends on Y and Z, and there are corresponding factories for Y and Z. So my code looks like this:

``` go
type X struct {
	Y Y
	Z Z
}

type Y struct {}
type Z struct {}

type YFactory interface {
	Create() (Y, error)
}

type ZFactory interface {
	Create() (Z, error)
}

type XFactory struct {
	yFactory YFactory
	zFactory ZFactory
}

// I hope this can be auto generated..
func (factory *XFactory) Create() (X, error) {
	y, err := factory.yFactory.Create()
	if err != nil {
		return err
	}
	z, err := factory.zFactory.Create()
	if err != nil {
		return err
	}
	return X{Y: y, Z: z}, nil
}

// wire.go
func InitializeXFactory() XFactory {
	wire.Build(NewXFactory, ProvideYFactory, ProvideZFactory)
	return nil
}
```

Wire allows me to inject customized Y and Z factory instances into the X factory, which is great. I hope `func (factory *XFactory) Create() (X, error)` can be also auto generated by wire. 

### Describe the solution you'd like

I hope I can do something like this for the `func (factory *XFactory) Create() (X, error)`:

``` go
func (factory *XFactory) Create() (X, error) {
	wire.Build(
		factory.YFactory.Create,
		factory.ZFactory.Create,
	)
	return X{}, nil
}
```

Please let me know if there is a better pattern for this purpose so that this feature request is not necessary.

### Describe alternatives you've considered

I can manually write the factory method while still taking advantage of wire to inject dependent factories.

**Comment by ijt** [comment]
Hi @zhengyangfeng00 , this looks like a question rather than a feature request or bug report, so it's best to move the discussion to https://groups.google.com/forum/#!forum/go-cloud. Closing the issue here but we'll be happy to discuss on the mailing list.

**Comment by vangent** [comment]
I'd like to take a look at this use case and see if we can update the documentation for it.

**Comment by vangent** [comment]
The particular things being asked about here are:
-- Why isn't it possible to use a non-top-level function as a provider? `factory.YFactory.Create` is not accepted by Wire.
-- Why isn't it possible to create a non-top-level injector function?

**Comment by vangent** [comment]
> Why isn't it possible to use a non-top-level function as a provider?

There is no good reason; it turns out there is already a feature request open for this:
https://github.com/google/go-cloud/issues/44

Please follow along there if you're interested.

> Why isn't it possible to create a non-top-level injector function?

`wire` could conceivably support this as well, but it's not obvious that there's a big benefit. With the pattern above, you've already done the hard work of identifying the exact dependencies for `X` by putting `YFactory` and `ZFactory` as fields in `XFactory`. You could also do a workaround using a top-level injector, something like this:

```
func (factory *XFactory) Create() (X, error) {
  return injectX(factory.yFactory, factory.zFactory)
}

func injectX(y YFactory, z ZFactory) (X, error)
  wire.Build(
    yFactory.Create,
    zFactory.Create,
  )
  return X{}, nil
}
```

This still depends on #44 being fixed. Right now, you would have to do something like:

```
func (factory *XFactory) Create() (X, error) {
  return injectX(factory.yFactory.Create, factory.zFactory.Create)
}

func injectX(y func() (Y, error), z func() (Z, error)) (X, error)
  wire.Build(y, z)
  return X{}, nil
}
```

Unfortunately, that doesn't work either, but I think it should; I opened #723, so you can follow along there if you're interested in that solution.

If you think that it would be useful for wire to create non-top-level injector functions, please open an issue specifically for that (and reference this issue for context).

Thanks, and thanks for trying out Wire!
